### ISSUE TITLE
Ionisation with 4th interpolation order

### ISSUE THREAD

**Issue by juliencelia** [SmileiPIC/Smilei#241]
Dear SMILEI Team,

I have an issue with a simulation with big target (200µm) of high density. I need to increase interpolation order to 4 but it bugs with ionization option (tunnel model). These two features are not compatible?

Thanks

Julien

**Comment by mccoys** [comment]
please indicate the bug

**Comment by Guilleaumes** [comment]
I think the "bug" is not a bug or an issue.

Simply, at the present day, in the file /src/Projector/Projector2D4Order.cpp for example, we can see, there is no current projection current for ionization with the 4th order. And so the code return an error.

```
// ---------------------------------------------------------------------------------------------------------------------
//! Project global current densities : ionization
// ---------------------------------------------------------------------------------------------------------------------
void  Projector2D4Order::ionizationCurrents( Field *Jx, Field *Jy, Field *Jz, Particles &particles, int ipart, LocalFields Jion )
{
    ERROR( "Projection of ionization current not yet defined for 2D 4th order" );
}
```

So 2 solutions for you : You can try to run a simulation with a current projection with a 4th order scheme. I write theses lines 2 years ago for my personal researches and test :

```
void  Projector2D4Order::operator() (Field* Jx, Field* Jy, Field* Jz, Particles &particles, int ipart, LocalFields Jion)
{
    Field2D* Jx2D  = static_cast<Field2D*>(Jx);
    Field2D* Jy2D  = static_cast<Field2D*>(Jy);
    Field2D* Jz2D  = static_cast<Field2D*>(Jz);

    //Declaration of local variables
    int ip, id, jp, jd;
    double xpn, xpmxip, xpmxip2, xpmxid, xpmxid2, xpmxip3, xpmxip4, xpmxid3, xpmxid4;
    double ypn, ypmyjp, ypmyjp2, ypmyjd, ypmyjd2, ypmyjp3, ypmyjp4, ypmyjd3, ypmyjd4;
    double Sxp[5], Sxd[5], Syp[5], Syd[5];
    
    // weighted currents
    double Jx_ion = Jion.x * particles.weight(ipart);
    double Jy_ion = Jion.y * particles.weight(ipart);
    double Jz_ion = Jion.z * particles.weight(ipart);

    //Locate particle on the grid
    xpn    = particles.position(0, ipart) * dx_inv_;  // normalized distance to the first node
    ypn    = particles.position(1, ipart) * dy_inv_;  // normalized distance to the first node

    // x-primal index
    ip      = round(xpn);                    // x-index of the central node
    xpmxip  = xpn - (double)ip;              // normalized distance to the nearest grid point
    xpmxip2 = xpmxip*xpmxip;                 // square of the normalized distance to the nearest grid point
    xpmxip3 = xpmxip2*xpmxip;
    xpmxip4 = xpmxip3*xpmxip;
    
    // x-dual index
    id      = round(xpn+0.5);                // x-index of the central node
    xpmxid  = xpn - (double)id + 0.5;        // normalized distance to the nearest grid point
    xpmxid2 = xpmxid*xpmxid;                 // square of the normalized distance to the nearest grid point
    xpmxid3 = xpmxid2*xpmxid;
    xpmxid4 = xpmxid3*xpmxid;

    // y-primal index
    jp      = round(ypn);                    // y-index of the central node
    ypmyjp  = ypn - (double)jp;              // normalized distance to the nearest grid point
    ypmyjp2 = ypmyjp*ypmyjp;                 // square of the normalized distance to the nearest grid point
    ypmyjp3 = ypmyjp2*ypmyjp;
    ypmyjp4 = ypmyjp3*ypmyjp;
    
    // y-dual index
    jd      = round(ypn+0.5);                // y-index of the central node
    ypmyjd  = ypn - (double)jd + 0.5;        // normalized distance to the nearest grid point
    ypmyjd2 = ypmyjd*ypmyjd;                 // square of the normalized distance to the nearest grid point
    ypmyjd3 = ypmyjd2*ypmyjd;
    ypmyjd4 = ypmyjd3*ypmyjd;

    Sxp[0] = dble_1_ov_384   - dble_1_ov_48  * xpmxip  + dble_1_ov_16 * xpmxip2 - dble_1_ov_12 * xpmxip3 + dble_1_ov_24 * xpmxip4;
    Sxp[1] = dble_19_ov_96   - dble_11_ov_24 * xpmxip  + dble_1_ov_4  * xpmxip2 + dble_1_ov_6  * xpmxip3 - dble_1_ov_6  * xpmxip4;
    Sxp[2] = dble_115_ov_192 - dble_5_ov_8   * xpmxip2 + dble_1_ov_4  * xpmxip4;
    Sxp[3] = dble_19_ov_96   + dble_11_ov_24 * xpmxip  + dble_1_ov_4  * xpmxip2 - dble_1_ov_6  * xpmxip3 - dble_1_ov_6  * xpmxip4;
    Sxp[4] = dble_1_ov_384   + dble_1_ov_48  * xpmxip  + dble_1_ov_16 * xpmxip2 + dble_1_ov_12 * xpmxip3 + dble_1_ov_24 * xpmxip4;

    Sxd[0] = dble_1_ov_384   - dble_1_ov_48  * xpmxid  + dble_1_ov_16 * xpmxid2 - dble_1_ov_12 * xpmxid3 + dble_1_ov_24 * xpmxid4;
    Sxd[1] = dble_19_ov_96   - dble_11_ov_24 * xpmxid  + dble_1_ov_4  * xpmxid2 + dble_1_ov_6  * xpmxid3 - dble_1_ov_6  * xpmxid4;
    Sxd[2] = dble_115_ov_192 - dble_5_ov_8   * xpmxid2 + dble_1_ov_4  * xpmxid4;
    Sxd[3] = dble_19_ov_96   + dble_11_ov_24 * xpmxid  + dble_1_ov_4  * xpmxid2 - dble_1_ov_6  * xpmxid3 - dble_1_ov_6  * xpmxid4;
    Sxd[4] = dble_1_ov_384   + dble_1_ov_48  * xpmxid  + dble_1_ov_16 * xpmxid2 + dble_1_ov_12 * xpmxid3 + dble_1_ov_24 * xpmxid4;

    Syp[0] = dble_1_ov_384   - dble_1_ov_48  * ypmyjp  + dble_1_ov_16 * ypmyjp2 - dble_1_ov_12 * ypmyjp3 + dble_1_ov_24 * ypmyjp4;
    Syp[1] = dble_19_ov_96   - dble_11_ov_24 * ypmyjp  + dble_1_ov_4  * ypmyjp2 + dble_1_ov_6  * ypmyjp3 - dble_1_ov_6  * ypmyjp4;
    Syp[2] = dble_115_ov_192 - dble_5_ov_8   * ypmyjp2 + dble_1_ov_4  * ypmyjp4;
    Syp[3] = dble_19_ov_96   + dble_11_ov_24 * ypmyjp  + dble_1_ov_4  * ypmyjp2 - dble_1_ov_6  * ypmyjp3 - dble_1_ov_6  * ypmyjp4;
    Syp[4] = dble_1_ov_384   + dble_1_ov_48  * ypmyjp  + dble_1_ov_16 * ypmyjp2 + dble_1_ov_12 * ypmyjp3 + dble_1_ov_24 * ypmyjp4;

    Syd[0] = dble_1_ov_384   - dble_1_ov_48  * ypmyjd  + dble_1_ov_16 * ypmyjd2 - dble_1_ov_12 * ypmyjd3 + dble_1_ov_24 * ypmyjd4;
    Syd[1] = dble_19_ov_96   - dble_11_ov_24 * ypmyjd  + dble_1_ov_4  * ypmyjd2 + dble_1_ov_6  * ypmyjd3 - dble_1_ov_6  * ypmyjd4;
    Syd[2] = dble_115_ov_192 - dble_5_ov_8   * ypmyjd2 + dble_1_ov_4  * ypmyjd4;
    Syd[3] = dble_19_ov_96   + dble_11_ov_24 * ypmyjd  + dble_1_ov_4  * ypmyjd2 - dble_1_ov_6  * ypmyjd3 - dble_1_ov_6  * ypmyjd4;
    Syd[4] = dble_1_ov_384   + dble_1_ov_48  * ypmyjd  + dble_1_ov_16 * ypmyjd2 + dble_1_ov_12 * ypmyjd3 + dble_1_ov_24 * ypmyjd4;

    ip  -= i_domain_begin;
    id  -= i_domain_begin;
    jp  -= j_domain_begin;
    jd  -= j_domain_begin;


for (unsigned int i=0 ; i<5 ; i++) {
        int iploc=ip+i-2;
        int idloc=id+i-2;
        for (unsigned int j=0 ; j<5 ; j++) {
            int jploc=jp+j-2;
            int jdloc=jd+j-2;
            // Jx^(d,p)
            (*Jx2D)(idloc,jploc) += Jx_ion * Sxd[i]*Syp[j];
            // Jy^(p,d)
            (*Jy2D)(iploc,jdloc) += Jy_ion * Sxp[i]*Syd[j];
            // Jz^(p,p)
            (*Jz2D)(iploc,jploc) += Jz_ion * Sxp[i]*Syp[j];
        }
    }//i
```
Or you can add a 2nd order projection and so copy paste the line of the /src/Projector/Projector2D2Order.cpp for example. The two solutions are not extensively tested, so its at your risk. Proceed somes test cases :)

Tell me if I'm wrong

**Comment by juliencelia** [comment]
Sorry for the wrong label. It is in fact a feature request;)

**Comment by MickaelGrech** [comment]
The solution by @Guilleaumes is correct; thanks for proposing it.
BTW @Guilleaumes next time do not hesitate to share your developments via a pull request.

I've started checking this more carefully throughout the code.
@juliencelia I'll get back to you asap, it should not be too complex as long as we do not try to vectorize this (which will come only later).
But if for now you need a quick fix, follow's @Guilleaumes advice.

**Comment by Guilleaumes** [comment]
I can do it for 1d and 2d if needed.

**Comment by MickaelGrech** [comment]
@Guilleaumes don't worry, I'm almost done.
Have to check the 3D one, and also implement validation scripts...
That last one is the most annoying.

**Comment by juliencelia** [comment]
Thank you @MickaelGrech and @Guilleaumes.

**Comment by juliencelia** [comment]
I am really not a developer but when I compile with @Guilleaumes advice compilation does not work:

Compiling src/Pusher/PusherPonderomotiveBorisV.cpp
src/Projector/Projector2D4Order.cpp(436): error: class "Projector2D4Order" has no member "operator()"
  void  Projector2D4Order::operator() (Field* Jx, Field* Jy, Field* Jz, Particles &particles, int ipart, LocalFields Jion)
                           ^

src/Projector/Projector2D4Order.cpp(454): error: identifier "dx_inv_" is undefined
      xpn    = particles.position(0, ipart) * dx_inv_;  // normalized distance to the first node
                                              ^

src/Projector/Projector2D4Order.cpp(455): error: identifier "dy_inv_" is undefined
      ypn    = particles.position(1, ipart) * dy_inv_;  // normalized distance to the first node
                                              ^

src/Projector/Projector2D4Order.cpp(485): error: identifier "dble_1_ov_384" is undefined
      Sxp[0] = dble_1_ov_384   - dble_1_ov_48  * xpmxip  + dble_1_ov_16 * xpmxip2 - dble_1_ov_12 * xpmxip3 + dble_1_ov_24 * xpmxip4;
               ^

src/Projector/Projector2D4Order.cpp(485): error: identifier "dble_1_ov_48" is undefined
      Sxp[0] = dble_1_ov_384   - dble_1_ov_48  * xpmxip  + dble_1_ov_16 * xpmxip2 - dble_1_ov_12 * xpmxip3 + dble_1_ov_24 * xpmxip4;
                                 ^

src/Projector/Projector2D4Order.cpp(485): error: identifier "dble_1_ov_16" is undefined
      Sxp[0] = dble_1_ov_384   - dble_1_ov_48  * xpmxip  + dble_1_ov_16 * xpmxip2 - dble_1_ov_12 * xpmxip3 + dble_1_ov_24 * xpmxip4;
                                                           ^

src/Projector/Projector2D4Order.cpp(485): error: identifier "dble_1_ov_12" is undefined
      Sxp[0] = dble_1_ov_384   - dble_1_ov_48  * xpmxip  + dble_1_ov_16 * xpmxip2 - dble_1_ov_12 * xpmxip3 + dble_1_ov_24 * xpmxip4;
                                                                                    ^

src/Projector/Projector2D4Order.cpp(485): error: identifier "dble_1_ov_24" is undefined
      Sxp[0] = dble_1_ov_384   - dble_1_ov_48  * xpmxip  + dble_1_ov_16 * xpmxip2 - dble_1_ov_12 * xpmxip3 + dble_1_ov_24 * xpmxip4;
                                                                                                             ^

src/Projector/Projector2D4Order.cpp(486): error: identifier "dble_19_ov_96" is undefined
      Sxp[1] = dble_19_ov_96   - dble_11_ov_24 * xpmxip  + dble_1_ov_4  * xpmxip2 + dble_1_ov_6  * xpmxip3 - dble_1_ov_6  * xpmxip4;
               ^

src/Projector/Projector2D4Order.cpp(486): error: identifier "dble_11_ov_24" is undefined
      Sxp[1] = dble_19_ov_96   - dble_11_ov_24 * xpmxip  + dble_1_ov_4  * xpmxip2 + dble_1_ov_6  * xpmxip3 - dble_1_ov_6  * xpmxip4;
                                 ^

src/Projector/Projector2D4Order.cpp(486): error: identifier "dble_1_ov_4" is undefined
      Sxp[1] = dble_19_ov_96   - dble_11_ov_24 * xpmxip  + dble_1_ov_4  * xpmxip2 + dble_1_ov_6  * xpmxip3 - dble_1_ov_6  * xpmxip4;
                                                           ^

src/Projector/Projector2D4Order.cpp(486): error: identifier "dble_1_ov_6" is undefined
      Sxp[1] = dble_19_ov_96   - dble_11_ov_24 * xpmxip  + dble_1_ov_4  * xpmxip2 + dble_1_ov_6  * xpmxip3 - dble_1_ov_6  * xpmxip4;
                                                                                    ^

src/Projector/Projector2D4Order.cpp(487): error: identifier "dble_115_ov_192" is undefined
      Sxp[2] = dble_115_ov_192 - dble_5_ov_8   * xpmxip2 + dble_1_ov_4  * xpmxip4;
               ^

src/Projector/Projector2D4Order.cpp(487): error: identifier "dble_5_ov_8" is undefined
      Sxp[2] = dble_115_ov_192 - dble_5_ov_8   * xpmxip2 + dble_1_ov_4  * xpmxip4;
                                 ^

src/Projector/Projector2D4Order.cpp(509): error: identifier "i_domain_begin" is undefined
      ip  -= i_domain_begin;
             ^

src/Projector/Projector2D4Order.cpp(511): error: identifier "j_domain_begin" is undefined
      jp  -= j_domain_begin;
             ^

src/Projector/Projector2D4Order.cpp(535): error: qualified name is not allowed
  void Projector2D4Order::currentsAndDensityWrapper( ElectroMagn *EMfields, Particles &particles, SmileiMPI *smpi, int istart, int iend, int ithread, bool diag_flag, bool is_spectral, int ispec, int icell, int ipart_ref )
       ^

src/Projector/Projector2D4Order.cpp(536): error: expected a ";"
  {
  ^

src/Projector/Projector2D4Order.cpp(566): warning #12: parsing restarts here after previous syntax error
  }
  ^

Compiling src/Pusher/PusherHigueraCary.cpp
compilation aborted for src/Projector/Projector2D4Order.cpp (code 2)
make: *** [build/src/Projector/Projector2D4Order.o] Error 2 

If that can help...

**Comment by Guilleaumes** [comment]
In fact, the code a share to you was a 2-years old piece of code. If you open the file of the 2D second order Projector you can see that the class hasn't the same member name today :

```
void Projector2D2Order::ionizationCurrents( Field *Jx, Field *Jy, Field *Jz, Particles &particles, int ipart, LocalFields Jion )
{
...
}
```
You have to replace 'operator' by 'ionizationCurrents'

The compilation works for me ;)

**Comment by MickaelGrech** [comment]
The implementation of 4th order projection for ionization current is available this last commit 65581263698c58418a597acc442e70fb21915255. This should answer to your feature request.
